[#18] 사용자 역할에 따른 접근 권한 설정

Hyung1Jung opened the pull request:
<!--- The issue this PR addresses (이 PR이 다루는 문제) -->
Fixes #18 
-  `로그인 하지 않은 사용자`, `로그인 한 사용자`, `관리자` 3개의 역할에 따른 페이지 접근 권한 설정
<!--- Detailed description of the change/feature (변경 / 기능에 대한 자세한 설명) -->
### 상세 내용
-  `로그인 하지 않은 사용자`, `로그인 한 사용자`, `관리자` 3개의 접근 권한에 따른 페이지 엑세스 기능

Hyung1Jung committed:
[#18] feat : 사용자 역할에 따른 접근 권한 설정

Hyung1Jung commented:
Admin 관련 기능은 우선 이 PR까지만 구현하도록 하고, 다음 PR부턴 User 관련 기능에 초점을 두고 개발을 진행할 예정입니다.

Hyung1Jung committed:
[#18] test : DisplayName 어노테이션 삭제 및 메서드명 수정

Hyung1Jung committed:
[#18] style(SessionLoginService) : 오탈자 수정

f-lab-dev commented on src/main/kotlin/com/flabedu/blackpostoffice/service/LoginService.kt:
```diff
@@ -6,4 +6,5 @@ interface LoginService {
 
     fun login(userLoginDto: UserLoginDto)
     fun loginCheck(userLoginDto: UserLoginDto)
+    fun setSessionAttribute(userLoginDto: UserLoginDto)
```
이렇게 하면 로그인서비스가 `세션`이라는 구체적인 방식에 의존적이게 되지 않나요?

Hyung1Jung replied:
네 맞습니다. 

LoginService를 둔 이유는 세션 로그인 방식 외에 `다른 로그인`에 관한 확장성을 고려해 인터페이스를 두었는데
인터페이스가 세션이라는 구체적인 방식에 의존하고 있습니다.

SessionLoginService에서 캡슐화 하는 것으로 수정해보도록 하겠습니다. 감사합니다.

f-lab-dev commented on src/main/kotlin/com/flabedu/blackpostoffice/exception/user/handler/UserExceptionHandler.kt:
```diff
@@ -34,4 +35,10 @@ class UserExceptionHandler {
+    @ExceptionHandler(AccessRejectedException::class)
+    fun handleAccessRejectedException(): ResponseEntity<String> {
+        logger.debug("이 페이지에 엑세스 하는데 필요한 권한이 존재하지 않습니다.")
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("이 페이지에 엑세스 하는데 필요한 권한이 존재하지 않습니다.")
```
body에 문자열을 보내는것보다 `ErrorResponse` 같은 공통 클래스를 하나 만들어서 json형식으로 보낼 수 있도록 하는게 좋지 않을까요?

Hyung1Jung replied:
네 맞습니다. 

ErrorResponse 같은 공통 클래스에 상수로 정의해서 재활용하는 것이 성능을 향상 시키는 방법이라고 생각합니다.

리팩토링 해보도록 하겠습니다. 감사합니다.

Hyung1Jung replied:
네. `ErrorResponse`와 같은 공통 클래스를 하나 만들어서 json 형식으로 리팩토링해보겠습니다.

원래는 `"message" : "이 페이지에 엑세스 하는데 필요한 권한이 존재하지 않습니다."`의 json 형식으로 보내주려고 생각했다가 ResponseEntity의 body를 이용하여 key값 없이 본문만 보냈습니다. 

이유는 `"message"`를 생략해도 응답 본문만 보고 응답 메세지를 충분히 이해할 수 있다고 생각하여 key값의 **리소스를 절약하여 성능 향상**을 위함이였는데요,

하지만 다시 생각해보니 그 성능이 미미할 뿐더러 json 형식을 지키는 것이 **더욱 명확하고**, json이 공식 규약인만큼 **json 형식을 지키는 것이 옳다고 생각**하여 지금은 말씀해주신대로 리팩토링 해보도록 하겠습니다.

f-lab-dev commented on src/main/kotlin/com/flabedu/blackpostoffice/util/annotation/LoginCheck.kt:
```diff
@@ -1,5 +1,9 @@
 @Target(AnnotationTarget.FUNCTION)
-annotation class LoginCheck()
+annotation class LoginCheck(
+    val authority: User.UserRole = User.UserRole.NONE
```
이렇게 표현을 보니 클래스명을 `UserRole`이라는 이름보다 그냥 `Role`로 지으면 `User.Role.None` 이렇게 표현할 수 있어서 더 간결해질 것 같아요~

Hyung1Jung replied:
UserRole은 User의 내부클래스이기 때문에  말씀해주신대로  `Role`로 표현해도 의미상 문제가 없고 훨씬 간결한 것 같습니다.  감사합니다.

f-lab-dev commented on src/main/kotlin/com/flabedu/blackpostoffice/util/interceptor/LoginCheckInterceptor.kt:
```diff
@@ -14,12 +16,28 @@ class LoginCheckInterceptor : HandlerInterceptor {
+            }
+        }
+    }
+
```
이렇게 검사해서 예외를 던져주는거는 `validate`라는 단어를 넣어서 메소드명을 지어주면 좋습니다~

Hyung1Jung replied:
네 리팩토링 해보겠습니다^^

Hyung1Jung committed:
[#18] refactor : 클래스명 수정 UserRole -> Role

Hyung1Jung committed:
[#18] refactor : Error response 형식 관련 수정 / body() -> JSON 형식

Hyung1Jung committed:
[#18] refactor : LoginService가 SessionLoginService에 의존하지 않도록 수정

Hyung1Jung committed:
[#18] refactor(GlobalLogger) : 전역적인 Logger 객체 사용을 위해 GlobalLogger 클래스 추가

Hyung1Jung committed:
[#18] refactor : 검사 관련 메서드 이름 수정

Hyung1Jung commented:
추후 `UserExceptionHandler`가 비대해지면 `UserLoginExceptionHandler`와 같이 더 분리할 예정입니다.

f-lab-dev commented on src/main/kotlin/com/flabedu/blackpostoffice/exception/user/AccessRejectedException.kt:
```diff
@@ -1,3 +1,13 @@
+
+class AccessRejectedException : RuntimeException(MESSAGE) {
+    companion object {
+        private const val MESSAGE = "이 페이지에 엑세스 하는데 필요한 권한이 존재하지 않습니다."
```
메세지를 외부에서 받는게 아니라 내부에서 이렇게 두는 이유가 무엇이실까요?

Hyung1Jung replied:
- 첫째는, **Admin**도 **User**의 범주 안에 있기 **Admin** 패키지를 따로 만들지 않을 것이라서  `AccessRejectedException`에서만 쓰일 것이라고 생각했습니다. 재사용하지 않는 선에선 내부에 정의해두는 건 더 읽기 쉬운 코드가 된다고 생각이 들었습니다.
- 둘째는, `UserExceptionHandler`가 비대해지거나, 아니면 추가로 공통 메세지를 정의한  `ResponseConstants` 같은 클래스가 생겼을때 그 클래스의 비대해지는 것을 줄이고자  생각을 하여 내부에 두었습니다.

두 곳 이상에서 쓰이는 메세지가 있다면 재사용을 고려해 `ResponseConstants`와 같은 클래스를 만들고 거기에 작성하려고 생각하고있지만,

현재는 `User`에서만 쓰일 것이라 생각하여 맨 위에서 말씀드린 두 가지의 장점을 살려보고자 하였습니다.

허나 다시 생각해보니 재활용 될 수도 있다는 것을 고려해 확장성과 `User`에서만 쓰인다고 하더라도, 각 `~~~Exception` 클래스를 찾아 들어가 유지보수를 하는 것보다, `ResponseConstants` 클래스를 정의하여 한 곳에 작성하여 바로 찾는 것이 유지보수하는 비용이 더 줄어들 것이라는 생각이 드네요.

`ResponseConstants` 클래스를 따로 두어 관리하도록 리팩토링해서 다시 올려보도록 하겠습니다.

Hyung1Jung committed:
[#18] refactor : ResponseConstants 클래스 생성하여 예외 메세지를 한 곳에서 관리

Hyung1Jung committed:
[#18] refactor(UserExceptionHandler) : 각 예외 클래스들의 로그 메시지를 핸들러 클래스로 이동

Hyung1Jung committed:
[#18] rename : 패키지명 수정 및 파일 위치 이동

Hyung1Jung committed:
[#16] 사용자별 접근 권한을 위해 회원가입시 역할 부여 (#17)

<!--- The issue this PR addresses (이 PR이 다루는 문제) -->
Fixes #16 
- 사용자별 접근을 위한 역할 (`로그인 하지 않은 사용자`, `사용자`, `관리자`) 이 나눠져 있지 않음.
- 메모리 사용 효율성, 간결성, 가독성을 유지보수 해야할 곳이 있음.

<!--- Detailed description of the change/feature (변경 / 기능에 대한 자세한 설명) -->
### 상세 내용
- 사용자별 접근을 위한 역할을 나눔
- 메모리 사용 효율성, 간결성, 가독성, 고려하여 코드 리팩토링

* [#16] feat : 사용자별 접근을 위한 역할 나눔 기능 추가

* [#16] refactor : 가독성 및 효율적인 메모리 사용을 위한 유지보수

* [#16] style(UserRole) : 들여쓰기 수정

* [#16] refactor : 접근 권한별 Enum name 수정

* [#16] refactor(UserDto) : 생성자 대신 companion object 사용

* [#16] test(UserControllerTest) : 회원가입 성공, 실패 코드 수정

* [#16] refactor(UserDto) : 생성자 대신 companion object 사용 취소

* [#16] refactor(UserDto) : 생성자 대신 companion object 사용 취소

* [#16] refactor : UserRole Enum Class -> UserDto에 내부 클래스로 이동

* [#16] remove(UserRole) : UserRole enum class 삭제

* [#16] refactor : UserRole 클래스를 UserDto에서 User의 내부클래스로 이동, createdAt도 이동

* [#16] refactor : 회원 정보 조회시 불필요하게 조회되는 컬럼 삭제

* [#16] refactor : @JvmOverloads 삭제 -> default 보조 생성자 추가

* [#16] refactor : private constructor() 제거

Hyung1Jung commented:
추가로 이 feature이 완료되면 로그아웃 feature도 같이 merge하도록 하겠습니다.

로그아웃 기능을 develop에 merge하는 것과 관련해서 

현재 feature의 commit이 로그아웃 feature에 붙어있기 때문에 rebase로 commit 삭제가 되지 않아서

현재 feature가 완료되면 merge후 브렌치 삭제후에 로그아웃 feature도 바로 merge 진행하도록 하겠습니다.

Hyung1Jung commented:
`로그아웃 피쳐`에 `현재 피쳐`의 커밋이 남아있기 때문에 rebase로 커밋을 삭제하는 것이 가능하지 않기에,
현재 피쳐가 승인이 되면,  `로그아웃 피쳐`와 함께 develop에 merge하도록 하겠습니다.

f-lab-dev commented on src/main/kotlin/com/flabedu/blackpostoffice/exception/user/DuplicateEmailException.kt:
```diff
@@ -1,13 +1,5 @@
-        logger.debug(MESSAGE)
-    }
-}
+class DuplicateEmailException : IllegalArgumentException(DUPLICATED_EMAIL)
```
이렇게 하더라도 역시 이 예외는 메세지를 커스터마이징 할 수 없어서 재사용성이 떨어지긴 하겠네요~

Hyung1Jung replied:
다시 고민해보니 메세지를 굳이 상수로 작성하는 것보다 커스터마이징하게 사용하는 것이 장점이 훨씬 많다고 느꼈습니다. 전체적으로 예외 메세지들을 커스터마이징 할 수 있게 리팩토링 해보았습니다.

`DuplicateEmailException` 같은 **이메일 중복에만 종속된** 예외 클래스가 아니라,

`DuplicateRequestException` **중복에 관한 모든 요청을 다룰 수 있는 추상적인 예외 클래스**로 만들어놓고,
`DuplicateRequestException(message: String)`처럼 **메세지를 갈아낄 수 있는 다형성**을 고려하였으므로 

- 예외 클래스들이 많이 추가되는 비용을 줄 일 수 있었고
- 많은 예외 클래스들을 알 필요 없이 추상적인 의미의 예외 클래스들 몇 개만 알고 메시지를 갈아끼는 다형성을 적용하면 되기 때문에 유지보수 하는 비용이 현저히 감소된다고 생각합니다.

혹시나 추후에 재사용 될 메세지가 있을 시에, 재사용 되는 메세지에 관해서만 상수로 관리 하는 것을 고려해보도록 하겠습니다.

재사용 되는 메세지에 관해 상수로 관리하는 것을 고려한 이유는,

재사용시 메세지에 똑같은 문자열을 복사해서 붙여넣기 식으로 재활용 할 수도 있겠지만, 
- 어떤 메세지를 붙여 넣어야 하는지 많은 클래스들을 뒤져봐야 하는 유지보수 비용이 들 뿐더러,
- 개발자의 실수로 단 한 글자 및 하나의 공백이라도 실수로 똑같이 쓰지 않게 된다면 **상수 풀의 낭비**가 발생할 수 있으므로

상수로 관리하도록 할 예정입니다.

리뷰 덕분에 좋은 고민 많이 한 것 같습니다. 감사합니다.

Hyung1Jung committed:
[#18] refactor : 커스터마이징하게 예외 핸들링

Hyung1Jung committed:
[#20] feat : 로그아웃 기능 추가 (#22)

로그아웃과 동시에 세션 무효화 기능 추가
로그아웃 기능 테스트코드 추가

Hyung1Jung committed:
[#18] docs : README 수정

Hyung1Jung committed:
[#18] style : 코드 포맷팅

Hyung1Jung committed:
[#18] refactor(SessionLoginCheck) : loginCheck 함수를 읽기 쉬운 코드로 리팩토링

Hyung1Jung committed:
[#18] refactor(UserExceptionHandler) : NullPointerException 방지를 위해서 !!을 ?.let 으로 수정

Hyung1Jung commented:
PR이 Closed 됬으므로,
현재 PR에  이어진, [다시 PR 요청한 링크](https://github.com/f-lab-edu/black-postoffice/pull/27) 첨부합니다.

Hyung1Jung committed:
[#18] refactor(LoginCheckInterceptor) : 코드 간소화 및 관심사 분리

Hyung1Jung committed:
[#18] refactor(LoginCheckInterceptor) : 변수명 수정

Hyung1Jung committed:
[#18] refactor(LoginCheckInterceptor) : 권한 체크 함수 수정

Hyung1Jung committed:
[#18] refactor(User, UserMapper) : userId 필드 추가

Hyung1Jung committed:
[#18] refactor : 접근 권한 검사, 로그인 요청 유효성 검사 함수 이름 수정

Hyung1Jung committed:
[#18] refactor : Logger 파일명 변경, 패키지 이동

Hyung1Jung committed:
[#18] refactor : 클래스 추적 가능한 Logger Interface 추가

Hyung1Jung committed:
[#18] rename : Logger 클래스 패키지 이동

Hyung1Jung committed:
[#18] refactor : logger.error -> log.debug로 변경

Hyung1Jung committed:
[#18] refactor : 불필요한 코드 삭제

Hyung1Jung committed:
[#18] docs : REDAME.md 수정

Hyung1Jung committed:
[#18] refactor : logger 관련 리팩토링

Hyung1Jung committed:
[#18] refactor : 들여쓰기 수정

Hyung1Jung committed:
Merge branch 'develop' into feature/18-access-by-role

# Conflicts:
#	src/main/kotlin/com/flabedu/blackpostoffice/domain/model/User.kt
#	src/main/kotlin/com/flabedu/blackpostoffice/service/LoginService.kt
#	src/main/kotlin/com/flabedu/blackpostoffice/util/interceptor/LoginCheckInterceptor.kt
#	src/main/resources/mapper/userMapper.xml
#	src/test/kotlin/com/flabedu/blackpostoffice/controller/UserLoginControllerTest.kt
